packages/opentelemetry-core/src/common/attributes.ts

```diff
@@ -29,7 +42,7 @@ export function sanitizeAttributes(attributes: unknown): SpanAttributes {
       continue;
     }
     if (!isAttributeValue(val)) {
-      diag.warn(`Invalid attribute value set for key: ${key}`);
+      out[key] = serializeNonPrimitiveAttributeValue(key, val);
```

Comment:
This change [violates the spec](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute):

> The attribute value is either:
> A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
> An array of primitive type values. The array MUST be homogeneous, i.e., it MUST NOT contain values of different types.

I am not sure if it is valid for opentracing or not (a relevant link to the spec can be nice here), but I vote against allowing it on all spans.
As long as the otel spec allows only primitive types, If the user wants to see the `JSON.strignfied` value as an attribute value, it's his responsibility to serialize it and call `setAttribute` with the resulting string.

Reply:
Do you want user to serialize the payload themselves before calling .log/.addEvent?

Reply:
Regarding `span.log`, since it's an opentracing api, I really don't know if it's valid or not.
If it is, then the place to add it is [here](https://github.com/open-telemetry/opentelemetry-js/blob/5005b0be18ad5f4b7a42abdd03e53a71e9cf7b5c/packages/opentelemetry-shim-opentracing/src/shim.ts#L290) so it's only affects opentracing.

What's not valid for sure is `span.setAttribute('foo', obj)` or `span.addEvent('bar', {'foo': obj})`.
If you think it's useful and should be permitted, the first step is adding it to the spec IMO.

Reply:
Looks like it is [allowed in opentracing](https://github.com/opentracing/specification/blob/master/specification.md#log-structured-data):
> Log structured data
> Required parameters
> - One or more key:value pairs, where the keys must be strings and the values may have any type at all

In this case, I support applying `JSON.stringify` only for `opentelemetry-shim-opentracing`, and not in `opentelemetry-core` where it will also affect otel spans.

Reply:
@blumamir 

I insist on doing it for `opentelemetry-shim-opentracing` because it should have backward compatibility with opentracing.Tracer, that allowed logging nested objects.

Should I create a new PR?

Reply:
> 

You can open a new PR or edit this one, whichever is easiest for you.
After reading your comments and the opentracing spec, I too think it should be fixed in `opentelemetry-shim-opentracing`. Thank you for working on this fix :)